exchanger/src/main/kotlin/com/d3/exchange/exchanger/ExchangerService.kt

```diff
@@ -0,0 +1,129 @@
+            }.map { command ->
+                command.transferAsset
+            }
+        }.map { exchangeCommands ->
+            exchangeCommands.forEach { exchangeCommand ->
+                performConversion(exchangeCommand)
+            }
+        }
+    }
+
+    private fun performConversion(transfer: Commands.TransferAsset) {
+        val sourceAsset = transfer.assetId
+        val targetAsset = transfer.description
+        val amount = transfer.amount
+        Result.of {
+            val relevantAmount = calculateRelevantAmount(sourceAsset, targetAsset, BigDecimal(amount))
+
+            ModelUtil.transferAssetIroha(
+                irohaConsumer,
+                exchangerAccountId,
+                transfer.srcAccountId,
+                targetAsset,
+                "Conversion from $sourceAsset to $targetAsset",
+                relevantAmount
+            )
+        }.fold({ logger.info { "Successfully converted $amount of $sourceAsset to $targetAsset." } },
+            {
+                logger.error("Exchanger error occurred. Performing rollback.", it)
+
+                ModelUtil.transferAssetIroha(
+                    irohaConsumer,
+                    exchangerAccountId,
+                    transfer.srcAccountId,
+                    sourceAsset,
+                    "Conversion rollback transaction",
+                    amount
+                )
+            })
+    }
+
+    private fun calculateRelevantAmount(from: String, to: String, amount: BigDecimal): String {
+        val fromAsset = BigDecimal(
+            getAccountAsset(queryAPI, irohaConsumer.creator, from).get()
+        ).minus(amount).toDouble()
+        val toAsset = BigDecimal(getAccountAsset(queryAPI, irohaConsumer.creator, to).get()).toDouble()
+        val amountMinusFee = amount.minus(amount.divide(FEE_PROPORTION)).toDouble()
+
+        val function = UnivariateFunction { x -> toAsset / (fromAsset * x) }
+        val precision = getAssetPrecision(queryAPI, to).fold({ it },
+            { throw AssetNotFoundException("Seems asset $to does not exist.") })
+
+        val integrate = integrator.integrate(EVALUATIONS, function, LOWER_BOUND, LOWER_BOUND + amountMinusFee)
+
+        if (integrate >= toAsset) {
+            throw TooMuchAssetVolumeException("Asset supplement exceeds the balance.")
+        }
+
+        return respectPrecision(integrate.toString(), precision)
+    }
+
+    private fun respectPrecision(rawValue: String, precision: Int): String {
+        val diff = rawValue.substringAfter('.').length - precision
+        if (diff >= 0) {
+            return rawValue
+        }
+        return rawValue.plus("0".repeat(diff * (-1)))
+    }
+
+    override fun close() {
+        chainListener.close()
+    }
+
+    companion object {
+        private const val EVALUATIONS = 1000
+        private const val LOWER_BOUND = 1.0
+        // 1% so far
+        private val FEE_PROPORTION = BigDecimal.valueOf(100)!!
```

Comment:
I think it's quite confusing to use proportion. I suggest using a ratio `amount.mult(FEE_RATIO)`, where `FEE_RATIO` is 0.01